### PR TITLE
[SPARK-32785][SQL][3.0] Interval with dangling parts should not result null

### PR DESCRIPTION
**Pull request by yaooqinn**
THIS PR brings https://github.com/apache/spark/pull/29635 to branch-3.0 and targets v3.0.2

### What changes were proposed in this pull request?

bugfix for incomplete interval values, e.g. interval '1', interval '1 day 2', currently these cases will result null, but actually we should fail them with IllegalArgumentsException

### Why are the changes needed?

correctness

### Does this PR introduce _any_ user-facing change?

yes, incomplete intervals will throw exception now

#### before
```
bin/spark-sql -S -e "select interval '1', interval '+', interval '1 day -'"

NULL NULL NULL
```
#### after

```
-- !query
select interval '1'
-- !query schema
struct<>
-- !query output
org.apache.spark.sql.catalyst.parser.ParseException

Cannot parse the INTERVAL value: 1(line 1, pos 7)

== SQL ==
select interval '1'
```

### How was this patch tested?

unit tests added

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan

File: docs/sql-migration-guide.md [hunk 1]
```diff
@@ -22,6 +22,10 @@ license: |
 * Table of contents
 {:toc}
 
+## Upgrading from Spark SQL 3.0.1 to 3.0.2
+
+  - In Spark 3.1, incomplete interval literals, e.g. `INTERVAL '1'`, `INTERVAL '1 DAY 2'` will fail with IllegalArgumentException. In Spark 3.0, they result `NULL`s.
```
Comment (cloud-fan) [inline docs/sql-migration-guide.md:27]: 3.0.2

**Review by cloud-fan — APPROVED** [review]

**Comment by yaooqinn** [comment]
retest this please

**Comment by cloud-fan** [comment]
github action passes, thanks, merging to 3.0!

**Pull request closed without merging**